src/main/java/org/wso2/lsp4intellij/utils/ApplicationUtils.java

```diff
@@ -56,4 +60,16 @@ static public void writeAction(Runnable runnable) {
     static public <T> T computableWriteAction(Computable<T> computable) {
         return ApplicationManager.getApplication().runWriteAction(computable);
     }
+
+    static public void invokeAfterPsiEvents(Runnable runnable) {
+        Runnable wrapper = () -> {
+            if(NoAccessDuringPsiEvents.isInsideEventProcessing()) {
+                invokeAfterPsiEvents(runnable);
+            } else {
+                runnable.run();
+            }
```

Comment:
```java
                Runnable wrapper = () -> {
                    while (NoAccessDuringPsiEvents.isInsideEventProcessing()) {
                        try {
                            Thread.sleep(500);
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                        }
                    }
                    runnable.run();
                };
```

IMO it would be better to have intervals between the checks as above. WDYT?

Reply:
I don’t think it’s a good idea to do like this. Because this will cause the event loop to wait stuck on the current event until the psi events are done which might cause a deadlock. The way I handle is by adding the event to tail always to give chance to finish the existing psi events in the event queue

Reply:
I don’t think it’s a good idea to do like this. Because this will cause the event loop to wait stuck on the current event until the psi events are done which might cause a deadlock. The way I handle is by adding the event to tail always to give chance to finish the existing psi events in the event queue

Reply:
I don’t think it’s a good idea to do like this. Because this will cause the event loop to wait stuck on the current event until the psi events are done which might cause a deadlock. The way I handle is by adding the event to tail always to give chance to finish the existing psi events in the event queue

Reply:
@gayanper Forgive me that I'm still not clear about your argument here but what I was worrying is the recursion depth as we are going recursively without any delay. Correct me If I'm wrong :)

Reply:
@NipunaRanasinghe I don’t think we have a recursion here. Because as soon as we push this runnable to invoke later the method will return. If we have a recursion actually this code should not work at all. Because until the current event pops out of the queue the remaining psi events won’t drain. Do you still think it can cause a recursion?

Reply:
@gayanper If I'm not mistaken you are invoking `invokeAfterPsiEvents(Runnable runnable);`  inside its own function body and that's what causes the recursion.

Reply:
@NipunaRanasinghe i think its little bit confusing how this code works, that might be the problem. Let me try to explain how i see it.

The first call comes into invokeAfterPsiEvents and the invocation is wrapped inside the new runnable. That runnable is pushed into the application event queue (AEQ). As soon as it it done the invokeAfterPsiEvents returns. So when the time comes to this runnable to run, it will check if still there are pending PSI events. If so we again call the invokeAfterPsiEvents with the orginal runnable instance we got. Now from the above we know that invokeAfterPsiEvents returns as soon as it pushes to AEQ. So i'm some what sure that there cannot be a recursion senario because we are using the second runnable wrapper and the AEQ. Do you still think there is a recursion ? if so please explain the steps how the recursion can occur because i might be still missing the view point you are having.

Reply:
Thanks for your explanation and I also think your concern is valid. It was just the implementation being somewhat tricky to understand at once, but now I also thinks that this should work as above. :)